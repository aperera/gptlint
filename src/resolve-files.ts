import fs from 'node:fs/promises'
import path from 'node:path'

import { globby } from 'globby'
import pMap from 'p-map'

import type * as types from './types.js'

export async function resolveFiles({
  config,
  cwd,
  concurrency = 16
}: {
  config: types.ResolvedLinterConfig
  cwd: string
  concurrency?: number
}) {
  const inputFiles = await globby(config.files, {
    gitignore: true,
    ignore: config.ignores,
    cwd
  })

  return readFiles(inputFiles, { concurrency })
}

export async function readFiles(
  filePaths: string[],
  {
    concurrency = 16,
    cwd = process.cwd()
  }: {
    concurrency?: number
    cwd?: string
  } = {}
): Promise<types.InputFile[]> {
  return pMap(
    filePaths,
    async (filePath) => {
      filePath = path.resolve(cwd, filePath)
      const content = await fs.readFile(filePath, { encoding: 'utf-8' })

      const fileRelativePath = path.relative(cwd, filePath)
      const fileName = path.basename(filePath)
      const ext = fileName.split('.').at(-1)?.toLowerCase() ?? ''
      const jsExtensions = new Set(['js', 'jsx', 'cjs', 'mjs'])
      const tsExtensions = new Set(['ts', 'tsx'])

      // TODO: improve filePath => language detection
      const language = jsExtensions.has(ext)
        ? 'javascript'
        : tsExtensions.has(ext)
        ? 'typescript'
        : 'unknown'

      return {
        filePath,
        fileRelativePath,
        fileName,
        language,
        content
      }
    },
    {
      concurrency
    }
  )
}
