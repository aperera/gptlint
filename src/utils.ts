import hashObject from 'hash-object'
import prettyMilliseconds from 'pretty-ms'

import type * as types from './types.js'
import { getLintDurationMs } from './lint-result.js'

export { default as slugify } from '@sindresorhus/slugify'
export { default as dedupe } from 'array-uniq'
export { default as assert } from 'tiny-invariant'

/**
 * From `obj`, create a new object that does not include `keys`.
 *
 * @example
 * ```
 * omit({ a: 1, b: 2, c: 3 }, 'a', 'c') // { b: 2 }
 * ```
 */
export const omit = <
  T extends Record<any, unknown>,
  K extends keyof T = keyof T
>(
  obj: T,
  ...keys: K[]
): Omit<T, K> => {
  const keysSet = new Set(keys)
  return Object.fromEntries(
    Object.entries(obj).filter(([k]) => !keysSet.has(k as any))
  ) as any
}

/**
 * From `obj`, create a new object that only includes `keys`.
 *
 * @example
 * ```
 * pick({ a: 1, b: 2, c: 3 }, 'a', 'c') // { a: 1, c: 3 }
 * ```
 */
export const pick = <
  T extends Record<any, unknown>,
  K extends keyof T = keyof T
>(
  obj: T,
  ...keys: K[]
): Pick<T, K> => {
  const keysSet = new Set(keys)
  return Object.fromEntries(
    Object.entries(obj).filter(([k]) => keysSet.has(k as any))
  ) as any
}

export function pruneUndefined<T extends Record<string, any>>(
  obj: T
): NonNullable<T> {
  return Object.fromEntries(
    Object.entries(obj).filter(([, value]) => value !== undefined)
  ) as NonNullable<T>
}

export function isValidRuleName(name: string): name is NonNullable<string> {
  if (!name) return false
  if (name.toLowerCase() !== name) return false

  const parts = name.split('/')
  if (parts.length === 2) {
    if (!/^@[a-z][\w-]*$/i.test(parts[0]!)) return false
    if (!/^[a-z][\w-]*$/i.test(parts[1]!)) return false
  } else if (!/^[a-z][\w-]*$/i.test(name)) return false

  return true
}

export function isValidRuleSetting(
  value: string
): value is types.LinterConfigRuleSetting {
  if (!value) return false
  if (value.toLowerCase() !== value) return false

  return value === 'off' || value === 'warn' || value === 'error'
}

export function trimMessage(
  message: string | undefined,
  { maxLength = 80 }: { maxLength?: number } = {}
): string {
  if (!message) return ''

  message = message.trim()
  if (message.length < maxLength) return message
  message = `${message.slice(0, maxLength - 3)}...`

  return message
}

export function createEvalStats(): types.EvalStats {
  return {
    numFiles: 0,
    numRules: 0,
    numUnexpectedErrors: 0,
    numFalseNegatives: 0,
    numFalsePositives: 0,
    numTrueNegatives: 0,
    numTruePositives: 0
  }
}

export function mergeEvalStats(
  evalStatsA: types.EvalStats,
  evalStatsB: types.EvalStats
): types.EvalStats {
  return {
    numFiles: evalStatsA.numFiles + evalStatsB.numFiles,
    numRules: evalStatsA.numRules + evalStatsB.numRules,
    numUnexpectedErrors:
      evalStatsA.numUnexpectedErrors + evalStatsB.numUnexpectedErrors,
    numFalseNegatives:
      evalStatsA.numFalseNegatives + evalStatsB.numFalseNegatives,
    numFalsePositives:
      evalStatsA.numFalsePositives + evalStatsB.numFalsePositives,
    numTrueNegatives: evalStatsA.numTrueNegatives + evalStatsB.numTrueNegatives,
    numTruePositives: evalStatsA.numTruePositives + evalStatsB.numTruePositives
  }
}

const knownSupportedCodeFileExtensions = new Set([
  'js',
  'ts',
  'mjs',
  'cjs',
  'jsx',
  'tsx',
  'py',
  'r',
  'rb',
  'html',
  'md',
  'c',
  'cpp',
  'cs',
  'h',
  'go',
  'java',
  'lisp',
  'lua',
  'css',
  'scss',
  'swift',
  'svg',
  'tcl',
  'txt',
  'csv',
  'tsv',
  'yaml',
  'json',
  'elm',
  'sh',
  'bash',
  'fish',
  'd',
  'dart'
])

const knownCodeFileMappings: Record<string, string> = {
  javascript: 'js',
  typescript: 'ts',
  markdown: 'md',
  python: 'py',
  text: 'txt'
}

export function inferBestPossibleCodeFileExtension(
  lang?: string | null,
  { fallbacks }: { fallbacks?: string[] } = {}
): string | undefined {
  function handleFallbacks() {
    if (fallbacks) {
      for (const fallback of fallbacks) {
        const alias = inferBestPossibleCodeFileExtension(fallback)
        if (alias) return alias
      }
    }

    // Fallback to no file extension
    return undefined
  }

  lang = lang?.trim().toLowerCase()

  if (!lang) return handleFallbacks()

  if (knownSupportedCodeFileExtensions.has(lang)) return lang

  const alias = knownCodeFileMappings[lang]
  if (alias) return alias

  return handleFallbacks()
}

export function getEnv(
  name: string,
  defaultValue?: string
): string | undefined {
  try {
    return (
      (typeof process !== 'undefined'
        ? // eslint-disable-next-line no-process-env
          process.env?.[name]
        : undefined) ?? defaultValue
    )
  } catch {
    return defaultValue
  }
}

export function logDebugConfig({
  files,
  rules,
  config
}: {
  files?: types.SourceFile[]
  rules: types.Rule[]
  config: types.ResolvedLinterConfig
}) {
  console.log(
    '\nlogging resolved config and then exiting because `debugConfig` is enabled'
  )

  const sanitizedConfig = pruneUndefined({
    ...config,
    llmOptions: pruneUndefined({
      ...config.llmOptions,
      apiKey: '<redacted>'
    })
  })

  console.log('\nconfig', JSON.stringify(sanitizedConfig, undefined, 2))
  console.log('\nrules', JSON.stringify(rules, undefined, 2))

  if (files) {
    console.log(
      '\ninput files',
      JSON.stringify(
        files.map((file) => file.fileRelativePath),
        undefined,
        2
      )
    )
  }
}

export function logLintResultStats({
  lintResult,
  config,
  prefix
}: {
  lintResult: types.LintResult
  config: types.ResolvedLinterConfig
  prefix?: string
}) {
  const lintDurationMs = getLintDurationMs(lintResult)
  const lintDuration = lintDurationMs
    ? prettyMilliseconds(lintDurationMs)
    : undefined

  console.log(
    `${prefix ?? ''}Linter stats; total cost $${(
      lintResult.totalCost / 100
    ).toFixed(2)}`,
    pruneUndefined({
      ...pick(config.llmOptions, 'model', 'weakModel'),
      ...pick(
        lintResult,
        'numModelCalls',
        'numModelCallsCached',
        'numPromptTokens',
        'numCompletionTokens',
        'numTotalTokens'
      ),
      lintDuration
    })
  )
}

export function logEvalStats({
  evalStats
}: // ruleToEvalStats
{
  evalStats: types.EvalStats
  // ruleToEvalStats: Record<string, types.EvalStats>
}) {
  const precision =
    evalStats.numTruePositives /
    (evalStats.numTruePositives + evalStats.numFalsePositives)
  const recall =
    evalStats.numTruePositives /
    (evalStats.numTruePositives + evalStats.numFalseNegatives)
  const f1Score = (2 * precision * recall) / (precision + recall)
  const extendedStats = {
    precision,
    recall,
    f1Score
  }

  console.log(`\nEval results`, { ...evalStats, ...extendedStats })
  return extendedStats
}

export function createCacheKey({
  file,
  rule,
  config
}: {
  file: types.SourceFile
  rule: types.Rule
  config: types.LinterConfig
}): string {
  // TODO: add linter major version to the cache key
  const cacheKeySource = {
    // Only keep the relative file path, content, and detected language
    file: pruneUndefined(pick(file, 'fileRelativePath', 'content', 'language')),

    // Remove rule fields which don't affect LLM logic
    rule: pruneUndefined(
      pick(
        rule,
        'name',
        'message',
        'desc',
        'positiveExamples',
        'negativeExamples',
        'languages'
      )
    ),

    // Ensure the cache key depends on how the LLM is parameterized
    llmOptions: pruneUndefined(
      pick(
        config.llmOptions ?? {},
        'model',
        'weakModel',
        'temperature',
        'apiBaseUrl'
      )
    )
  }

  return hashObject(cacheKeySource)
}

export function createPromiseWithResolvers<
  T = unknown
>(): PromiseWithResolvers<T> {
  let resolve: (value: T | PromiseLike<T>) => void
  let reject: (reason: any) => void

  const promise = new Promise<T>((res, rej) => {
    resolve = res
    reject = rej
  })

  return { promise, resolve: resolve!, reject: reject! }
}
