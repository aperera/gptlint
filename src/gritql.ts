import { execa } from 'execa'
import plur from 'plur'
import which from 'which'

import type * as types from './types.js'
import * as constants from './constants.js'
import { createLintResult } from './lint-result.js'
import { assert } from './utils.js'

/**
 * @see https://github.com/getgrit/gritql
 */
export namespace grit {
  export interface Match {
    __typename: Typename
    messages: any[]
    variables: Variable[]
    sourceFile: string
    ranges: Range[]
    level?: number
    message?: string
    debug?: string
  }

  export type Typename = 'Match' | 'AnalysisLog' | 'AllDone'

  export interface Range {
    start: FileOffset
    end: FileOffset
    startByte: number
    endByte: number
  }

  export interface FileOffset {
    line: number
    column: number
  }

  export interface Variable {
    name: VariableName
    scopedName: ScopedName
    ranges: Range[]
  }

  export type VariableName =
    | '$absolute_filename'
    | '$all_imports'
    | '$anchor'
    | '$body'
    | '$filename'
    | '$GLOBAL_IMPORTED_NAMES'
    | '$GLOBAL_IMPORTED_SOURCES'
    | '$h'
    | '$imported_names'
    | '$imports'
    | '$joined_imported_names'
    | '$match'
    | '$name'
    | '$new_files'
    | '$p'
    | '$program'
    | '$source'
    | '$statements'
    | string

  export type ScopedName = string
}

/**
 * Resolves a GritQL pattern against the given source files and then resolves
 * the matches to construct partial source files containing only the matched
 * portions.
 */
export async function resolveGritQLPattern(
  pattern: string,
  {
    files,
    numLinesContext = constants.gritNumLinesContext
  }: {
    files: types.SourceFile[]
    // Number of lines of context to include around each matching range
    numLinesContext?: number
  }
): Promise<Map<string, types.PartialSourceFile>> {
  if (!(await hasGrit())) {
    return new Map()
  }

  const matches = await applyGritQLPattern(pattern, { files })
  return resolveGritQLMatches(matches, { files, numLinesContext })
}

/**
 * Runs `grit apply --dry-run --jsonl` on the given `pattern` and `files`.
 *
 * Returns an array of matches.
 */
export async function applyGritQLPattern(
  pattern: string,
  {
    files
  }: {
    files: types.SourceFile[]
  }
): Promise<grit.Match[]> {
  const gritBinary = await whichGritBinary()
  if (!gritBinary) {
    throw new Error(`Could not find 'grit' binary in $PATH`)
  }

  const paths = files.map((file) => file.filePath)

  const res = await execa(gritBinary, [
    'apply',
    '--dry-run',
    '--jsonl',
    pattern,
    ...paths
  ])

  // Parse the jsonl output for matches
  const lines = res.stdout
    .split('\n')
    .map((line) => line.trim())
    .filter(Boolean)
  const gritMatches = lines
    .map((line) => {
      try {
        const potentialMatch = JSON.parse(line) as grit.Match
        if (potentialMatch?.__typename !== 'Match') {
          return undefined
        }

        const { debug: _, ...match } = potentialMatch
        return match
      } catch {
        return undefined
      }
    })
    .filter(Boolean)

  return gritMatches
}

export async function hasGrit(): Promise<boolean> {
  const gritBinary = await whichGritBinary()
  return !!gritBinary
}

export async function whichGritBinary(): Promise<string | null> {
  // TODO: make sure this matches `node_modules/.bin/grit`
  return which('grit', { nothrow: true })
}

/**
 * Takes an array of GritQL `matches` and the `files` they're matched against,
 * and returns an array of partial source files containing only the matched
 * portions of the given files.
 */
export function resolveGritQLMatches(
  matches: grit.Match[],
  {
    files,
    numLinesContext = constants.gritNumLinesContext
  }: {
    files: types.SourceFile[]
    // Number of lines of context to include around each matching range
    numLinesContext?: number
  }
): Map<string, types.PartialSourceFile> {
  const matchesByFilePath = new Map<string, grit.Match[]>()

  // Group matches by absolute file path
  for (const match of matches) {
    // TODO: verify if this is using absolute or relative paths
    if (!matchesByFilePath.has(match.sourceFile)) {
      matchesByFilePath.set(match.sourceFile, [])
    }
    matchesByFilePath.get(match.sourceFile)!.push(match)
  }

  const partialFilesByFilePath = new Map<string, types.PartialSourceFile>()

  // Group partial files by absolute file path
  for (const file of files) {
    const { filePath } = file

    if (!partialFilesByFilePath.has(filePath)) {
      partialFilesByFilePath.set(filePath, {
        ...file,
        ranges: [],
        partialContent: ''
      })
    }
  }

  // Concat all the ranges for each file
  for (const [filePath, matches] of matchesByFilePath.entries()) {
    const partialFile = partialFilesByFilePath.get(filePath)!
    assert(partialFile)

    const matchRanges = matches.flatMap(
      (m) => m.variables.find((v) => v.name === '$match')?.ranges ?? m.ranges
    )

    partialFile.ranges = partialFile.ranges
      .concat(matchRanges)
      .sort((a, b) => a.start.line - b.start.line)
  }

  // Aggregate the partial content for each file based on the matched ranges
  for (const partialFile of partialFilesByFilePath.values()) {
    const lines = partialFile.content.split('\n')
    const { ranges } = partialFile

    let partialContentLines: string[] = []
    let maxLine = -1

    for (const range of ranges) {
      const startLine = Math.max(
        Math.max(0, maxLine - 1),
        range.start.line - 1 - numLinesContext
      )
      const endLine = range.end.line + numLinesContext
      if (startLine >= endLine) continue

      maxLine = Math.max(endLine, maxLine)
      const partialLines = lines.slice(startLine, endLine)
      partialContentLines = partialContentLines.concat(partialLines)
    }

    partialFile.partialContent = partialContentLines
      .map((line) => line.trim())
      .filter(Boolean)
      .join('\n')
  }

  return partialFilesByFilePath
}

export async function preProcessFileWithGrit({
  file,
  files,
  rule,
  config,
  ruleNameToPartialSourceFileMap = new Map()
}: {
  file: types.SourceFile | types.PartialSourceFile
  rule: types.Rule
  config: types.ResolvedLinterConfig
  files?: (types.SourceFile | types.PartialSourceFile)[]
  ruleNameToPartialSourceFileMap?: Map<
    string,
    Promise<Map<string, types.PartialSourceFile>>
  >
}): Promise<types.LintResult | undefined> {
  if (!rule.gritql || config.linterOptions.noGrit) {
    return
  }

  if (!ruleNameToPartialSourceFileMap.has(rule.name)) {
    const partialSourceFileMapP = resolveGritQLPattern(rule.gritql, {
      files: files ?? [file],
      numLinesContext: rule.gritqlNumLinesContext
    })
    ruleNameToPartialSourceFileMap.set(rule.name, partialSourceFileMapP)

    const partialSourceFileMap = await partialSourceFileMapP
    if (config.linterOptions.debugGrit) {
      console.log(
        `gritql pattern "${rule.gritql}" matches:\n\n${[
          ...partialSourceFileMap.values()
        ]
          .map(
            (f) =>
              `  ${f.fileRelativePath} found ${f.ranges?.length || 0} ${plur('match', f.ranges?.length || 0)}`
          )
          .join('\n\n')}\n\n`
      )
    }
  }

  const partialSourceFileMap = await ruleNameToPartialSourceFileMap.get(
    rule.name
  )!

  const partialSourceFile = partialSourceFileMap.get(file.filePath)!
  assert(partialSourceFile)

  file.ranges = partialSourceFile.ranges
  file.partialContent = partialSourceFile.partialContent.trim()

  if (!file.ranges.length || !file.partialContent) {
    return createLintResult({
      skipped: true,
      skipReason: 'grit-pattern',
      skipDetail: 'no gritql matches'
    })
  }
}
