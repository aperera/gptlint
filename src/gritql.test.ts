import path from 'node:path'

import { expect, test } from 'vitest'

import { resolveGritQLPattern } from './gritql.js'
import { readSourceFiles } from './resolve-files.js'
import { dirname } from './utils.js'

test('', async () => {
  const files = await readSourceFiles([
    path.resolve(dirname(), './constants.ts')
  ])
  const filesMap = await resolveGritQLPattern('identifier', {
    files
  })

  expect(filesMap.size).toEqual(1)
  const file = filesMap.get(files[0]?.filePath!)!
  expect(file).toBeTruthy()
  expect(file.partialContent).toBeTruthy()
  expect(file.partialContent).toMatchSnapshot()
})
