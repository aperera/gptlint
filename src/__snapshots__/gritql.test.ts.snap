// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[` 1`] = `
"export const minSourceFileSizeBytes = 0
export const maxSourceFileSizeBytes = 200 * 1024 // 200kb

export const maxSourceFileNumLines = 10_000
export const maxSourceFileLineLength = 1024

export const gritNumLinesContext = 3"
`;
